File: spring-data-graphql-commons/src/main/java/io/github/wickedev/graphql/repository/GraphQLDataLoaderByIdRepository.java
package io.github.wickedev.graphql.repository;

import graphql.schema.DataFetchingEnvironment;
import io.github.wickedev.graphql.interfases.Node;
import io.github.wickedev.graphql.types.ID;
import org.springframework.data.repository.NoRepositoryBean;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * In Kotlin findById, due to the return type of T or T?,
 * the user cannot freely write the desired type due to strict classification,
 * so it is written in Java.
 */
@NoRepositoryBean
public interface GraphQLDataLoaderByIdRepository<T extends Node> extends GraphQLDataLoaderRepository<T> {
    CompletableFuture<T> findById(ID id, DataFetchingEnvironment env);

    CompletableFuture<List<T>> findAllById(Iterable<ID> ids, DataFetchingEnvironment env);

    CompletableFuture<Boolean> existsById(ID id, DataFetchingEnvironment env);
}

File: spring-data-graphql-commons/src/main/kotlin/io/github/wickedev/graphql/repository/GraphQLDataLoaderByIdRepository.kt
package io.github.wickedev.graphql.repository

import graphql.schema.DataFetchingEnvironment
import io.github.wickedev.graphql.interfases.Node
import io.github.wickedev.graphql.types.ID
import org.springframework.data.repository.NoRepositoryBean
import java.util.concurrent.CompletableFuture

@NoRepositoryBean
interface GraphQLDataLoaderByIdRepository<T : Node> : GraphQLDataLoaderRepository<T> {

    fun findById(id: ID, env: DataFetchingEnvironment): CompletableFuture<T?>

    fun findAllById(ids: Iterable<ID>, env: DataFetchingEnvironment): CompletableFuture<List<T>>

    fun existsById(id: ID, env: DataFetchingEnvironment): CompletableFuture<Boolean>
}